### PR TITLE
Ucx

### PR DESCRIPTION
**Pull request by boeschf**
- ucx stuff
- initialization
- threads
- factory

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -40,33 +55,57 @@ target_include_directories(ghexlib INTERFACE
 target_link_libraries(ghexlib INTERFACE GridTools::gridtools MPI::MPI_CXX)
```
Comment (lukasm91) [inline CMakeLists.txt]: here you should link against `UCP::libucp` or `PMIx::libpmix` if this is enabled

File: CMakeLists.txt [hunk 2]
```diff
@@ -40,33 +55,57 @@ target_include_directories(ghexlib INTERFACE
...
-target_link_libraries(gtest_main_bench MPI::MPI_CXX GridTools::gridtools GTest::gtest)
-
-add_library(gtest_main_bench_mt ./utils/gtest_main_bench.cpp)
-target_link_libraries(gtest_main_bench_mt MPI::MPI_CXX GridTools::gridtools GTest::gtest)
-target_compile_definitions(gtest_main_bench_mt PRIVATE GHEX_BENCHMARKS_USE_MULTI_THREADED_MPI)
+enable_testing()
 
-add_subdirectory(tests)
-add_subdirectory(benchmarks)
+if (GHEX_BUILD_TESTS OR GHEX_BUILD_BENCHMARKS)
```
Comment (lukasm91) [inline CMakeLists.txt]: ususally called `BUILD_TESTING` (or if you really want `GHEX_BUILD_TESTING`)

File: CMakeLists.txt [hunk 3]
```diff
@@ -40,33 +55,57 @@ target_include_directories(ghexlib INTERFACE
...
+    # https://github.com/google/googletest/issues/2429
+    add_library(GTest::gtest ALIAS gtest)
+  endif()
+  
+  set(GTEST_LIB_SOURCES ${googletest_SOURCE_DIR}/googletest/src/gtest-all.cc)
+  set(GTEST_INCLUDE_DIRS ${googletest_SOURCE_DIR}/googletest/include ${googletest_SOURCE_DIR}/googletest/src ${googletest_SOURCE_DIR}/googletest/)
+  
+  if (GHEX_BUILD_TESTS)
+    add_library(gtest_main_mt ${GTEST_LIB_SOURCES} ./utils/gtest_main.cpp )
+    target_include_directories(gtest_main_mt PRIVATE ${CMAKE_SOURCE_DIR}/include ${GTEST_INCLUDE_DIRS})
```
Comment (lukasm91) [inline CMakeLists.txt]: you shouldn't do this (link against `GTest::gtest`, see old code)

File: benchmarks/CMakeLists.txt [hunk 4]
```diff
@@ -25,38 +25,39 @@ set(_benchmarks gcl_test_halo_exchange_3D_generic_full comm_2_test_halo_exchange
...
-  target_link_libraries(${_t} MPI::MPI_CXX GridTools::gridtools gtest_main_bench ghexlib)
+    add_executable(${_t} ${_t}.cpp)
+    target_link_libraries(${_t} gtest_main_bench)
 
-  add_executable(${_t}_1_pattern ${_t}.cpp)
-  target_compile_definitions(${_t}_1_pattern PUBLIC GHEX_1_PATTERN_BENCHMARK)
-  target_link_libraries(${_t}_1_pattern MPI::MPI_CXX GridTools::gridtools gtest_main_bench ghexlib)
+    add_executable(${_t}_1_pattern ${_t}.cpp)
+    target_compile_definitions(${_t}_1_pattern PUBLIC GHEX_1_PATTERN_BENCHMARK GCL_MPI) 
+    target_link_libraries(${_t}_1_pattern gtest_main_bench)
```
Comment (lukasm91) [inline benchmarks/CMakeLists.txt]: incorrect indentation (I usually use two spaces, which seems to be the convention now; I would really stick to it).

File: CMakeLists.txt (hunk 3 above)
Comment (boeschf) [reply CMakeLists.txt]: This won't work - it segfaults on my machine. This is the only way so far I could get it to work

File: include/ghex/pattern.hpp [hunk 5]
```diff
@@ -11,9 +11,10 @@
 #ifndef INCLUDED_GHEX_PATTERN_HPP
 #define INCLUDED_GHEX_PATTERN_HPP
 
-#include "./transport_layer/mpi/setup.hpp"
-#include "./transport_layer/mpi/communicator.hpp"
+//#include "./transport_layer/mpi/setup.hpp"
```
Comment (mbianco) [inline include/ghex/pattern.hpp]:
```suggestion
```

File: include/ghex/pattern.hpp [hunk 6]
```diff
@@ -11,9 +11,10 @@
...
+//#include "./transport_layer/mpi/communicator.hpp"
```
Comment (mbianco) [inline include/ghex/pattern.hpp]:
```suggestion
```

File: include/ghex/transport_layer/mpi/communicator_base.hpp [hunk 7]
```diff
@@ -63,7 +64,13 @@ namespace gridtools{
                     /** @return size of communicator group*/
                     inline size_type size() const noexcept { return m_size; }
 
-                    void barrier() const { MPI_Barrier(*m_comm); }
+                    static void initialize() {}
+		            static void finalize() {}
```
Comment (mbianco) [inline include/ghex/transport_layer/mpi/communicator_base.hpp]:
```suggestion
		      static void finalize() {}
```

File: include/ghex/transport_layer/raw_shared_message_buffer.hpp [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+
+namespace gridtools {
+    namespace ghex {
+        namespace tl {
+
+	    /* A struct that keeps the message with the refcount.
+	       The shared message buffer keeps a pointer to a dynamically
+	       allocated refcounted_message. That pointer can be shared
+	       between many messages. 
+	     */
```
Comment (mbianco) [inline include/ghex/transport_layer/raw_shared_message_buffer.hpp]: This has implications on thread safety, right? Either this or next comment should clarify. We have many pieces around and it's difficult for me to keep all them in context. We should prepare a presentation for the others in order to put everything in perspective (I would need it to glue together stuff)

File: include/ghex/transport_layer/shared_message_buffer.hpp [hunk 9]
```diff
@@ -38,7 +38,7 @@ namespace gridtools {
     
                 static constexpr bool can_be_shared = true;
 
-            private: // members
+            //private: // members
```
Comment (mbianco) [inline include/ghex/transport_layer/shared_message_buffer.hpp]:
```suggestion
```
?

**Pull request merged**